### ISSUE TITLE
Regex router not applied

### ISSUE THREAD

**Issue by gunnarmorling** [confluentinc/kafka-connect-elasticsearch#246]
I'm trying to use the regex router to lower-case the index name for a given topic. But the router isn't applied, causing an exception as the original topic name isn't a valid index name:

```
Invalid index name [dbserver1.inventory.Hike], must be lowercase\",\"index_uuid\":\"_na_\",\"index\":\"dbserver1.inventory.Hike\"} at 
io.confluent.connect.elasticsearch.jest.JestElasticsearchClient.createIndices(JestElasticsearchClient.java:238) at 
io.confluent.connect.elasticsearch.ElasticsearchWriter.createIndicesForTopics(ElasticsearchWriter.java:330) at 
io.confluent.connect.elasticsearch.ElasticsearchSinkTask.open(ElasticsearchSinkTask.java:157)
```

That's the connector config:

```
{
    "name": "elastic-sink",
    "config": {
        "connector.class": "io.confluent.connect.elasticsearch.ElasticsearchSinkConnector",
        "tasks.max": "1",
        "topics": "Customers",
        "connection.url": "http://elastic:9200",
        "transforms": "unwrap,key,route",
        "transforms.unwrap.type": "io.debezium.transforms.UnwrapFromEnvelope",
        "transforms.key.type": "org.apache.kafka.connect.transforms.ExtractField$Key",
        "transforms.key.field": "id",
        "transforms.route.type": "org.apache.kafka.connect.transforms.RegexRouter",
        "transforms.route.regex": "Customers",
        "transforms.route.replacement": "customers",
        "key.ignore": "false",
        "type.name": "customer"
    }
}
```

I have also an executable reproducer, if you'd like to take a look: https://github.com/gunnarmorling/debezium-examples/tree/kafka-connect-elasticsearch-99/unwrap-smt

Check out that branch and follow the steps in https://github.com/gunnarmorling/debezium-examples/blob/kafka-connect-elasticsearch-99/unwrap-smt/reproducer.md. Registering the sink connector fails due to an upper-case index name, which should have been transformed by means of the regex router. This is with Kafka 2.0 and ES sink 5.0.0.

**Comment by purbon** [comment]
Hi, thanks for reporting this @gunnarmorling , certainly this is a bug, however a quick test where I was able to verify the problem, I am not sure this is a problem of transformations. From my understanding the transformations are used to change event per event, however the plugin try to create the indexes at startup time, this is what I understand from:

```java
[2018-10-16 13:11:22,494] ERROR WorkerSinkTask{id=elasticsearch-sink-0} Task threw an uncaught and unrecoverable exception (org.apache.kafka.connect.runtime.WorkerTask:177)
org.apache.kafka.connect.errors.ConnectException: Could not create index 'Customers': {"root_cause":[{"type":"invalid_index_name_exception","reason":"Invalid index name [Customers], must be lowercase","index_uuid":"_na_","index":"Customers"}],"type":"invalid_index_name_exception","reason":"Invalid index name [Customers], must be lowercase","index_uuid":"_na_","index":"Customers"}
	at io.confluent.connect.elasticsearch.jest.JestElasticsearchClient.createIndices(JestElasticsearchClient.java:238)
	at io.confluent.connect.elasticsearch.ElasticsearchWriter.createIndicesForTopics(ElasticsearchWriter.java:330)
	at io.confluent.connect.elasticsearch.ElasticsearchSinkTask.open(ElasticsearchSinkTask.java:157)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.openPartitions(WorkerSinkTask.java:612)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.access$1100(WorkerSinkTask.java:69)
	at org.apache.kafka.connect.runtime.WorkerSinkTask$HandleRebalance.onPartitionsAssigned(WorkerSinkTask.java:672)
	at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.onJoinComplete(ConsumerCoordinator.java:283)
	at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.joinGroupIfNeeded(AbstractCoordinator.java:422)
	at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.ensureActiveGroup(AbstractCoordinator.java:352)
	at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.ensureActiveGroup(AbstractCoordinator.java:337)
	at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.poll(ConsumerCoordinator.java:333)
	at org.apache.kafka.clients.consumer.KafkaConsumer.updateAssignmentMetadataIfNeeded(KafkaConsumer.java:1218)
	at org.apache.kafka.clients.consumer.KafkaConsumer.poll(KafkaConsumer.java:1181)
	at org.apache.kafka.clients.consumer.KafkaConsumer.poll(KafkaConsumer.java:1115)
```

I am not sure transformations are the good way to lowercase. 

In my opinion the best way to do that is to force all indexes to be created lowercase as this is the official way elasticsearch is handling that, see https://github.com/elastic/elasticsearch/issues/29420 for more details.

What do the others @confluentinc/connect think about it?

**Comment by jbotuck** [comment]
Not sure I understand. It appears that transformations are the recommended way to go, that is why topic.index.map was deprecated

**Comment by gunnarmorling** [comment]
I don't think it's a good idea to alter the index name automatically (e.g. to lower-case it), as it may be quite surprising to the user. Could you perhaps create the index upon first arrival of a source record from a given topic? Then you'd automatically apply whatever name transformations have been configured.

If that's not viable, I'd question the decision to deprecate the old mapping option and suggest to un-deprecate it, as it can safely be applied eagerly to create the index(es) as well as when handling individual source records.

**Comment by egorzaremba** [comment]
Do you have any news?

**Comment by babadofar** [comment]
Fixed in 5.2.0, perhaps https://docs.confluent.io/5.2.0/connect/kafka-connect-elasticsearch/changelog.html#version-5-2-0
https://github.com/confluentinc/kafka-connect-elasticsearch/pull/275

**Comment by levzem** [comment]
fixed by #251

**Comment by jbotuck** [comment]
Should this issue really be closed? Is the regex router applied for index names? If not, should we undeprecate topic.index.map?

**Comment by levzem** [comment]
@jbotuck the issue is fixed #251 and #275 addresses the problem of creating the indexes at start up by introducing a new config, thus deprecating `topic.index.map`.
